File: src/tests/get_audio_links.py
import json
import re
from typing import List
from urllib.parse import quote, urljoin, urlparse

import requests
from bs4 import BeautifulSoup


def get_voh_audio_links(url: str) -> List[str]:
    """漢聲廣播電臺"""
    response = requests.get(url)
    soup = BeautifulSoup(response.text, "html.parser")
    audio_tag = soup.find("div", class_="Audio").find("audio")
    audio_src = audio_tag["src"]
    parsed_url = urlparse(url)
    base_url = f"{parsed_url.scheme}://{parsed_url.netloc}"
    audio_url = urljoin(base_url, audio_src)
    audio_url = quote(audio_url, safe=":/?=&")
    return [audio_url]


def get_ner_audio_links(url: str) -> List[str]:
    """國立教育廣播電臺"""
    parsed_url = urlparse(url)
    news_id = parsed_url.path.split("/")[-1]
    api_news_url = urljoin(
        parsed_url._replace(path="/").geturl(), f"/api/news/{news_id}"
    )
    response = requests.get(api_news_url)
    audio_id = response.json()["audio"]
    audio_url = urljoin(
        parsed_url._replace(path="/").geturl(), f"/api/audio/{audio_id}"
    )
    audio_url = audio_url + ".mp3"
    return [audio_url]


def get_tradio_audio_links(url: str) -> List[str]:
    """聽台北 tradio"""
    parsed_url = urlparse(url)
    program_id = parsed_url.path.split("/")[-1]
    response = requests.get(url)
    soup = BeautifulSoup(response.text, "html.parser")

    # Track down the audio URL
    data = json.loads(soup.find("script", id="__NUXT_DATA__").string)
    idx = data[2][f"program-{program_id}"]
    idx = data[idx]["player_url"]
    player_url = data[idx]

    response2 = requests.get(player_url)
    soup2 = BeautifulSoup(response2.text, "html.parser")
    script_text = soup2.find_all("script", src="")
    audio_base_url = re.search(
        r"src:\s*'([^']+)'", script_text[0].string + script_text[1].string
    ).group(1)
    audio_base_url = re.sub(r"playlist.m3u8.+", "", audio_base_url)
    audio_urls = []
    for i in range(350):
        audio_urls.append(f"{audio_base_url}/media_{i+1}.aac")
    return audio_urls

File: src/getAudioLinks.js
export async function getAudioLinks(url) {
  const parsedUrl = new URL(url)
  const domain = parsedUrl.hostname

  if (domain.includes('voh.com.tw')) {
    return await getVohAudioLinks(url)
  } else if (domain.includes('ner.gov.tw')) {
    return await getNerAudioLinks(url)
  } else if (domain.includes('tradio.gov.taipei')) {
    return await getTradioAudioLinks(url)
  } else {
    throw new Error(`Unsupported domain: ${domain}`)
  }
}

// Example usage:
// getAudioLinks('https://www.voh.com.tw/TW/Playback/ugC_Playback.aspx?PID=168&D=20240917').then(console.log);
// getAudioLinks('https://www.ner.gov.tw/news/66e258d1425f580023b5f80b').then(console.log);
// getAudioLinks('https://tradio.gov.taipei/channel/MjEyOA/MjIwNTg0').then(console.log);

async function edgeFetch(url) {
  const response = await fetch('/api/edgeFetch', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ url })
  })

  if (!response.ok) {
    throw new Error(`Error: ${response.statusText}`)
  }

  const text = await response.text()
  return text
}

// 漢聲廣播電臺
// https://www.voh.com.tw/TW/Playback/ugC_Playback.aspx?PID=168&D=20240917
async function getVohAudioLinks(url) {
  const text = await edgeFetch(url)
  const parser = new DOMParser()
  const doc = parser.parseFromString(text, 'text/html')
  const audioTag = doc.querySelector('div.Audio audio')
  const audioSrc = audioTag.getAttribute('src')
  const baseUrl = new URL(url).origin
  const audioUrl = new URL(audioSrc, baseUrl).href
  return [audioUrl]
}

// 國立教育廣播電臺
// https://www.ner.gov.tw/news/66e258d1425f580023b5f80b
async function getNerAudioLinks(url) {
  const parsedUrl = new URL(url)
  const newsId = parsedUrl.pathname.split('/').pop()
  const apiNewsUrl = new URL(`/api/news/${newsId}`, parsedUrl.origin).href
  const text = await edgeFetch(apiNewsUrl)
  const data = JSON.parse(text)
  const audioId = data.audio
  const audioUrl = new URL(`/api/audio/${audioId}.mp3`, parsedUrl.origin).href
  return [audioUrl]
}

// 聽台北 Tradio
// https://tradio.gov.taipei/channel/MjEyOA/MjIwNTg0
async function getTradioAudioLinks(url) {
  const parsedUrl = new URL(url)
  const text = await edgeFetch(parsedUrl)
  const parser = new DOMParser()
  // Get the json data
  const doc = parser.parseFromString(text, 'text/html')
  const scriptTag = doc.querySelector('#__NUXT_DATA__')
  // Get the player url from the json data
  const data = JSON.parse(scriptTag.textContent)
  // console.log(`getTradioAudioLinks: data: ${data}`)
  const programId = parsedUrl.pathname.split('/').pop()
  let idx = 2
  console.log(`getTradioAudioLinks: idx: ${idx}, ${JSON.stringify(data[idx])}`)
  idx = data[idx][`program-${programId}`]
  console.log(`getTradioAudioLinks: idx: ${idx}, ${JSON.stringify(data[idx])}`)
  idx = data[idx+1]['player_url']
  console.log(`getTradioAudioLinks: idx: ${idx}, ${JSON.stringify(data[idx])}`)
  const playerUrl = data[idx]
  // Get the audio urls from the player
  console.log(`getTradioAudioLinks: playerUrl: ${playerUrl}`)
  const text2 = await edgeFetch(new URL(playerUrl))
  const doc2 = parser.parseFromString(text2, 'text/html')
  const scriptText = Array.from(doc2.querySelectorAll('script:not([src])'))
    .map((script) => script.textContent)
    .join('')
  const audioBaseUrl = scriptText.match(/src:\s*'([^']+)'/)[1].replace(/playlist.m3u8.+/, '')
  const audioUrls = Array.from({ length: 350 }, (_, i) => `${audioBaseUrl}media_${i + 1}.aac`)
  return audioUrls
}
